Code/GraphMol/FileParsers/MaeMolSupplier.cpp

```diff
@@ -10,243 +10,424 @@
+void set_mol_properties(RWMol &mol, const mae::Block &ct_block) {
+  for (const auto &prop : ct_block.getProperties<std::string>()) {
+    if (prop.first == mae::CT_TITLE) {
+      mol.setProp(common_properties::_Name, prop.second);
+    } else if (prop.first.find(mae::CT_CHIRALITY_PROP_PREFIX) == 0 ||
+               prop.first.find(mae::CT_PSEUDOCHIRALITY_PROP_PREFIX) == 0) {
+      parseChiralityLabel(mol, prop.second);
+    } else if (prop.first.find(mae::CT_EZ_PROP_PREFIX) == 0) {
+      parseStereoBondLabel(mol, prop.second);
+    } else {
+      mol.setProp(prop.first, prop.second);
     }
+  }
+  for (const auto &prop : ct_block.getProperties<double>()) {
+    mol.setProp(prop.first, prop.second);
+  }
+  for (const auto &prop : ct_block.getProperties<int>()) {
+    mol.setProp(prop.first, prop.second);
+  }
+  for (const auto &prop : ct_block.getProperties<mae::BoolProperty>()) {
+    mol.setProp(prop.first, static_cast<bool>(prop.second));
+  }
+}
 
-    if(pdb_info.m_insertion_code &&
-            pdb_info.m_insertion_code->isDefined(atom_num))
-    {
-        rd_info->setInsertionCode(pdb_info.m_insertion_code->at(atom_num));
+//! Set atom properties. Some of these have already been parsed to construct the
+//! Atom object, and should be skipped. Also, atom properties may be undefined
+//! for an atom, and should also be skipped for that atom.
+void set_atom_properties(Atom &atom, const mae::IndexedBlock &atom_block,
+                         size_t i) {
+  for (const auto &prop : atom_block.getProperties<std::string>()) {
+    if (prop.first == PDB_ATOM_NAME || prop.first == PDB_RESIDUE_NAME ||
+        prop.first == PDB_CHAIN_NAME || prop.first == PDB_INSERTION_CODE) {
+      // PDB information is parsed separately.
+      continue;
+    } else if (!prop.second->isDefined(i)) {
+      continue;
     }
 
-    if(pdb_info.m_resnum && pdb_info.m_resnum->isDefined(atom_num))
-    {
-        rd_info->setResidueNumber(pdb_info.m_resnum->at(atom_num));
+    atom.setProp(prop.first, prop.second->at(i));
+  }
+
+  for (const auto &prop : atom_block.getProperties<double>()) {
+    if (prop.first == mae::ATOM_X_COORD || prop.first == mae::ATOM_Y_COORD ||
+        // Coordinates are used in defining a conformation, and should not be
+        // set on the atom.
+        prop.first == mae::ATOM_Z_COORD) {
+      continue;
+    } else if (prop.first == PDB_OCCUPANCY || prop.first == PDB_TFACTOR) {
+      // PDB information is parsed separately.
+      continue;
+    } else if (!prop.second->isDefined(i)) {
+      continue;
     }
 
-    if(pdb_info.m_occupancy && pdb_info.m_occupancy->isDefined(atom_num))
-    {
-        rd_info->setOccupancy(pdb_info.m_occupancy->at(atom_num));
+    atom.setProp(prop.first, prop.second->at(i));
+  }
+  for (const auto &prop : atom_block.getProperties<int>()) {
+    if (prop.first == mae::ATOM_ATOMIC_NUM) {
+      // Atomic number was already used in the creation of the atom
```

Comment:
Should we delete the property when we consume it?

Reply:
I don't really have a preference about that. I don't think it is going to be a big performance improvement, so maybe we can just leave it be, and do it together with the other change about setting one property for all atoms at a time.